plugins/contents/fps_contents/fileid.py

```diff
@@ -0,0 +1,175 @@
+
+                    if change == Change.deleted:
+                        async with db.execute(
+                            "SELECT * FROM fileids WHERE path = ?", (changed_path,)
+                        ) as cursor:
+                            async for _ in cursor:
+                                break
+                            else:
+                                # path is not indexed, ignore
+                                continue
+                        # path is indexed
+                        await maybe_rename(db, changed_path, deleted_paths, added_paths, False)
+                    elif change == Change.added:
+                        await maybe_rename(db, changed_path, added_paths, deleted_paths, True)
+
+                for path in deleted_paths + added_paths:
+                    await db.execute("DELETE FROM fileids WHERE path = ?", (path,))
+                await db.commit()
+
+                for change in changes:
+                    changed_path = change[1]
+                    # get relative path
+                    changed_path = str(Path(changed_path).relative_to(Path().absolute()))
+                    for watcher in self.watchers.get(changed_path, []):
+                        watcher.notify(change)
+
+    def watch(self, path: str) -> Watcher:
+        watcher = Watcher(path)
+        if path not in self.watchers:
+            self.watchers[path] = watchers = []  # type: ignore
+        watchers.append(watcher)
+        return watcher
+
+
+async def get_mtime(path, db) -> Optional[float]:
+    if db:
+        async with db.execute("SELECT * FROM fileids WHERE path = ?", (path,)) as cursor:
+            async for _, _, mtime in cursor:
+                return mtime
+            # deleted file is not in database, shouldn't happen
+            return None
+    try:
+        mtime = (await AsyncPath(path).stat()).st_mtime
+    except FileNotFoundError:
+        return None
+    return mtime
+
+
+async def maybe_rename(
+    db, changed_path: str, changed_paths: List[str], other_paths: List[str], is_added_path
+) -> None:
+    # check if the same file was added/deleted, this would be a rename
+    db_or_fs1, db_or_fs2 = db, None
+    if is_added_path:
+        db_or_fs1, db_or_fs2 = db_or_fs2, db_or_fs1
+    mtime1 = await get_mtime(changed_path, db_or_fs1)
+    if mtime1 is None:
+        return
+    for other_path in other_paths:
+        mtime2 = await get_mtime(other_path, db_or_fs2)
+        if mtime1 == mtime2:
+            # same files, according to modification times
+            path1, path2 = changed_path, other_path
+            if is_added_path:
+                path1, path2 = path2, path1
+            await db.execute("UPDATE fileids SET path = REPLACE(path, ?, ?)", (path1, path2))
+            other_paths.remove(other_path)
+            return
+    changed_paths.append(changed_path)
+
+
+FILE_ID_MANAGER: Optional[FileIdManager] = None
+
+
+def get_file_id_manager() -> FileIdManager:
```

Comment:
Looks like this is a singleton pattern. It might be better to add a `get_instance()` _classmethod_ on `FileIdManager` so that it's self-contained and would remove the need for a global (since the single instance would reside within the class).

Doesn't the `db_path` constructor argument need to be present here?